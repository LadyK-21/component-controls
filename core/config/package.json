{
  "name": "@component-controls/config",
  "version": "3.11.0",
  "description": "Component controls configuration reading utilities",
  "main": "dist/index.js",
  "module": "dist/index.esm.js",
  "typings": "dist/index.d.ts",
  "files": [
    "dist/",
    "package.json",
    "README.md"
  ],
  "scripts": {
    "build": "yarn cross-env NODE_ENV=production rollup -c",
    "dev": "yarn cross-env NODE_ENV=development rollup -cw",
    "docs": "ts-md",
    "fix": "yarn lint --fix",
    "lint": "yarn eslint . --ext mdx,ts,tsx",
    "prepare": "yarn build",
    "test": "yarn jest"
  },
  "homepage": "https://github.com/ccontrols/component-controls",
  "bugs": {
    "url": "https://github.com/ccontrols/component-controls/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/ccontrols/component-controls.git",
    "directory": "core/config"
  },
  "license": "MIT",
  "dependencies": {
    "@component-controls/core": "^3.11.0",
    "@component-controls/instrument": "^3.11.0",
    "@component-controls/render": "^3.11.0",
    "glob": "^7.1.6",
    "glob-base": "^0.3.0",
    "micromatch": "^4.0.2",
    "typescript": "^4.0.5",
    "yargs": "^15.3.1"
  },
  "devDependencies": {
    "@component-controls/ts-markdown-docs": "^3.8.2",
    "@types/faker": "^4.1.9",
    "@types/glob": "^7.1.3",
    "@types/glob-base": "^0.3.0",
    "@types/jest": "^26.0.10",
    "@types/micromatch": "^4.0.1",
    "cross-env": "^5.2.1",
    "eslint": "^6.5.1",
    "jest": "^26.4.2"
  },
  "esm": {
    "cjs": true
  },
  "browser": {
    "fs": false
  },
  "jest": {
    "testEnvironment": "node",
    "roots": [
      "."
    ]
  },
  "publishConfig": {
    "access": "public"
  },
  "authors": [
    "Atanas Stoyanov"
  ],
  "gitHead": "e30d62f101e104711a16261fce04785d58cac1eb"
}
