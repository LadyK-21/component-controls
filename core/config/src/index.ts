import * as path from 'path';
import * as fs from 'fs';

import { sync as globSync } from 'glob';
import globBase from 'glob-base';
import { makeRe } from 'micromatch';

import yargs from 'yargs';
import { BuildConfiguration } from '@component-controls/core';

export const buildConfigFileNames = [
  'buildtime.js',
  'build.js',
  'main.js',
  'doczrc.js',
];

export const optionsFileNames = [
  'runtime.js',
  'options.js',
  'runtime.ts',
  'runtime.tsx',
  'options.ts',
  'options.tsx',
];
export interface ConfigurationResult {
  config: BuildConfiguration;
  configPath: string;
  optionsFilePath?: string;
}
/**
 * return the configuration folder from command-line parameters
 * command line accepts -c/ -config parameter for config path
 * the config file is assumed named main.js/main.ts
 * @param baseFolder project folder to start the search with
 */
export const getConfigurationArg = (
  args: string[] = process.argv,
): string | undefined => {
  const argv = yargs(args)
    .option('config', {
      alias: 'c',
      description: 'configuration folder',
      type: 'string',
    })
    .help()
    .alias('help', 'h').argv;
  return argv.config;
};

/**
 *  given a base project folder and a configuration folder, returns the configuration file
 *
 * @param baseFolder project folder to start the search with
 * @param configFolder folder where the configuration file is located
 * @param args optional arguments
 */
export const loadConfiguration = (
  baseFolder: string,
  configFolder?: string,
  args?: string[],
  defaultConfigPath?: string,
): ConfigurationResult | undefined => {
  const folder = configFolder ?? getConfigurationArg(args) ?? defaultConfigPath;
  const configPath = folder ? path.resolve(baseFolder, folder) : baseFolder;
  const hasConfigFolder = fs.existsSync(configPath);
  if (!hasConfigFolder) {
    console.warn('configuration folder not found', configPath);
  }
  const allFiles = hasConfigFolder ? fs.readdirSync(configPath) : [];
  const buildConfigFile = allFiles.find(file =>
    buildConfigFileNames.includes(file.toLowerCase()),
  );
  const optionsFile = allFiles.find(file =>
    optionsFileNames.includes(file.toLowerCase()),
  );
  if (buildConfigFile) {
    const buildPath = path.resolve(configPath, buildConfigFile);
    let config = require('esm')(module)(buildPath);
    if (
      !config ||
      (typeof config === 'object' && Object.keys(config).length === 0)
    ) {
      config = require(buildPath);
    }
    return {
      config: config.default || config,
      optionsFilePath: optionsFile
        ? path.resolve(configPath, optionsFile)
        : undefined,
      configPath,
    };
  }
  return undefined;
};

//fix for sb5 issue handling glob
const fixGlob = (golbExpr: string): string => golbExpr.replace('.(', '.@(');

/**
 * find the story files out of a configuration file
 * using glob for the regex file search
 */
export const extractDocuments = ({
  config,
  configPath,
}: ConfigurationResult): string[] | undefined => {
  const stories = config ? config.stories || config.files : undefined;
  const files = typeof stories === 'string' ? [stories] : stories;
  const documents = files
    ? files.reduce((acc: string[], storyRg: string) => {
        const matches = globSync(path.resolve(configPath, fixGlob(storyRg)));
        if (!matches.length) {
          throw new Error(`${storyRg} did not match any files`);
        }
        const ignore = config.ignore || [];
        const files = matches.filter(filename => {
          const basename = path.basename(filename).toLowerCase();
          return !ignore.includes(basename);
        });
        return [...acc, ...files];
      }, [])
    : undefined;
  return documents;
};

/**
 * from the glob list of documents, extract require.context array of props
 */
export interface RequireContextProps {
  directory: string;
  useSubdirectories: boolean;
  regExp: RegExp;
}
/**
 * convert glob patters from config file into require.context objects
 */
export const configRequireContext = ({
  config,
  configPath,
}: ConfigurationResult): RequireContextProps[] | undefined => {
  const stories = config ? config.stories || config.files : undefined;
  const files = typeof stories === 'string' ? [stories] : stories;

  const contexts = files
    ? files.reduce((acc: RequireContextProps[], storyRg: string) => {
        const base = globBase(fixGlob(storyRg));
        const useSubdirectories = base.glob.startsWith('**');
        const glob = useSubdirectories ? base.glob.substr(3) : base.glob;
        const regExp = base.isGlob
          ? RegExp(makeRe(glob).source.substr(1))
          : new RegExp(glob);
        return [
          ...acc,
          {
            directory: path.resolve(configPath, base.base),
            useSubdirectories,
            regExp,
          },
        ];
      }, [])
    : undefined;
  return contexts;
};

/**
 * merge a configuration passed through cli or tools, with the build configration from the config path
 */
export const mergeBuildConfiguration = (
  config: BuildConfiguration,
): BuildConfiguration => {
  const { configPath } = config;
  const buildConfig = loadConfiguration(process.cwd(), configPath);
  return { ...buildConfig?.config, ...config };
};
