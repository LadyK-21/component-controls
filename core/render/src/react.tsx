/* eslint-disable react/display-name */
import { createElement } from 'react';
import {
  StoryRenderFn,
  getControlValues,
  deepMerge,
  FrameworkRenderFn,
} from '@component-controls/core';

export const render: FrameworkRenderFn = (story, doc, options: any = {}) => {
  if (!story) {
    throw new Error(`Invalid story`);
  }
  if (!doc) {
    throw new Error(`Invalid doc`);
  }
  const controls = story.controls;
  const values = getControlValues(controls);
  //parameters added to avoid bug in SB6 rc that assumes parameters exist
  const context = {
    story,
    doc,
    controls,
    ...options,
  };
  const { decorators: globalDecorators = [] } = options;
  const { decorators: storyDecorators = [] } = story;
  const decorators = deepMerge<StoryRenderFn[]>(
    globalDecorators,
    storyDecorators,
  );
  const sortedDecorators = decorators.reverse();
  let renderFn = story.renderFn;
  for (let i = 0; i < sortedDecorators.length; i += 1) {
    const decorator = sortedDecorators[i];
    const childFn = renderFn;
    const nextRenderFn = (_: any, nextContext: any) =>
      (childFn as StoryRenderFn)(values, { ...context, ...nextContext });
    renderFn = () =>
      decorator(nextRenderFn, {
        ...context,
        renderFn: nextRenderFn,
      });
  }
  return typeof renderFn === 'function'
    ? createElement(() => (renderFn as StoryRenderFn)(values, context))
    : createElement('div', 'invalid render function');
};
