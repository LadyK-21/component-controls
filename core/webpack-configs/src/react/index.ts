import path from 'path';
import { RuleSetLoader } from 'webpack';
import ExtractCssPlugin from 'extract-css-chunks-webpack-plugin';
import OptimizeCssAssetsWebpackPlugin from 'optimize-css-assets-webpack-plugin';

import {
  PresetType,
  BuildProps,
  defCssFileName,
  customLoaderOptions,
} from '@component-controls/core';
import { findUpFile } from '@component-controls/core/node-utils';

export const react: PresetType = (options: BuildProps) => {
  const isProd = process.env.NODE_ENV === 'production';
  const cssLoaders: RuleSetLoader[] = [];
  const postcssOptions = customLoaderOptions(options, 'postcss-loader', {});
  const postCssOptionsFile = findUpFile(process.cwd(), 'postcss.config.js');
  const hasPostCss = Object.keys(postcssOptions).length || postCssOptionsFile;
  if (hasPostCss && !((postcssOptions as any).disable === true)) {
    cssLoaders.push({
      loader: 'postcss-loader',
      options: {
        postcssOptions:
          typeof postCssOptionsFile === 'string'
            ? require(postCssOptionsFile)
            : undefined,
        sourceMap: true,
      },
    });
  }
  const result: PresetType = {
    plugins: [
      new ExtractCssPlugin({
        filename: options.cssFileName || defCssFileName,
      }),
    ],
    optimization: { minimizer: [] },
    performance: { hints: false },
    module: {
      rules: [
        {
          test: /\.(js|jsx)$/,
          exclude: [/node_modules/],
          use: [
            {
              loader: 'babel-loader',
              options: {
                presets: [
                  [
                    '@babel/preset-env',
                    {
                      targets: {
                        browsers: ['last 5 versions', 'ie >= 9'],
                        node: 'current',
                      },
                      modules: 'commonjs',
                      useBuiltIns: 'usage',
                      corejs: 3,
                    },
                  ],
                  ['@babel/preset-react', { runtime: 'automatic' }],
                ],
              },
            },
          ],
        },
        {
          test: /\.(ts|tsx)$/,
          exclude: [/node_modules/],
          use: [
            {
              loader: 'babel-loader',
              options: {
                presets: [['react-app', { flow: false, typescript: true }]],
              },
            },
          ],
        },
        {
          test: /\.(md|mdx)$/i,
          exclude: [/node_modules/],
          loader: 'babel-loader',
          options: {
            presets: [
              [
                '@babel/preset-env',
                {
                  targets: {
                    browsers: ['last 5 versions', 'ie >= 9'],
                    node: 'current',
                  },
                  modules: 'commonjs',
                  useBuiltIns: 'usage',
                  corejs: 3,
                },
              ],
              '@babel/preset-react',
            ],
          },
        },
        {
          test: /\.(eot|md|svg|ico|jpg|jpeg|png|gif|ttf|woff|woff2|pdf|mp4|web|wav|mp3|m4a|aac|oga)$/i,
          exclude: [/node_modules/],
          loader: 'url-loader',
          options: customLoaderOptions(options, 'url-loader', {
            limit: 25000,
            name: '[name].[ext]',
            publicPath: '/static',
            outputPath: path.relative(
              options?.distFolder || process.cwd(),
              path.resolve(options?.staticFolder || process.cwd()),
            ),
          }),
        },
        {
          test: /\.(woff(2)?|ttf|eot|svg)(\?v=\d+\.\d+\.\d+)?$/,
          use: [
            {
              loader: 'file-loader',
              options: customLoaderOptions(options, 'file-loader', {
                name: '[name].[ext]',
                outputPath: path.relative(
                  options?.distFolder || process.cwd(),
                  path.resolve(options?.staticFolder || process.cwd()),
                ),
              }),
            },
          ],
        },
        {
          test: /\.(css|sass|scss|less)$/i,
          use: [
            // Creates `style` nodes from JS strings
            // will export to a consolidated css file
            ExtractCssPlugin.loader,
            {
              // Translates CSS into CommonJS
              loader: 'css-loader',
              options: customLoaderOptions(options, 'css-loader', {
                sourceMap: true,
              }),
            },
            ...cssLoaders,
            {
              // Compiles Sass to CSS
              loader: 'sass-loader',
              options: customLoaderOptions(options, 'sass-loader', {
                sourceMap: true,
              }),
            },
            {
              // Compiles less to CSS
              loader: 'less-loader',
              options: customLoaderOptions(options, 'less-loader', {
                sourceMap: true,
              }),
            },
          ],
        },
        {
          test: /\.txt$/i,
          use: [
            {
              loader: 'raw-loader',
              options: customLoaderOptions(options, 'raw-loader', {}),
            },
          ],
        },

        {
          test: /\.(md|mdx)$/i,
          exclude: [/node_modules/],
          loader: '@component-controls/loader/loader',
          enforce: 'pre',
          options: {
            mdx: {
              transformMDX: true,
            },
          },
        },
        {
          test: /\.(story|stories|doc|docs).(js|jsx|ts|tsx)$/,
          loader: '@component-controls/loader/loader',
          exclude: [/node_modules/],
          enforce: 'pre',
        },
      ],
    },
    resolve: {
      extensions: ['.ts', '.tsx', '.js', '.jsx'],
    },
  };
  if (isProd) {
    result.optimization = {
      minimizer: [
        new OptimizeCssAssetsWebpackPlugin({
          cssProcessorOptions: {
            discardComments: { removeAll: true },
          },
        }),
      ],
    };
  }
  return result;
};
