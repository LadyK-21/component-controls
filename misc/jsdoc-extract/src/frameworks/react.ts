import { FrameworkPlugin, JSDocType, JSDocTypes, JSImports } from '../utils';
import { walkProps } from '../utils/resolve-props';

const reactImport = (
  names: JSDocType[],
  imports: JSImports,
): JSDocType | undefined => {
  const reactImports = imports.filter(i => i.module === 'react');
  const foundReact = names?.find(({ name }) => {
    const nameParts = name?.split('.');
    if (nameParts) {
      const namedImport = nameParts.pop();
      const defaultImport = nameParts.length ? nameParts[0] : null;
      if (namedImport) {
        return reactImports.some(
          i => i.name === defaultImport || i.namedImports?.[namedImport],
        );
      }
    }
    return false;
  });
  return foundReact;
};
const reactClassComponent = (
  component: JSDocType,
  imports: JSImports,
): JSDocType | undefined => {
  const classes = component.inherits?.filter(i => i.type === 'class' && i.name);
  if (classes) {
    return reactImport(classes, imports);
  }
  return undefined;
};

const reactFunctionComponent = (
  component: JSDocType,
  imports: JSImports,
): JSDocType | undefined => {
  if (component.type === 'function') {
    if (component.value) {
      return reactImport([component.value], imports);
    }
    if (component.parameters?.length) {
      return {
        parameters: component.parameters,
      };
    }
  }

  return undefined;
};
const reactComponent = (
  component: JSDocType,
  imports: JSImports,
): JSDocType | undefined => {
  return (
    reactClassComponent(component, imports) ||
    reactFunctionComponent(component, imports)
  );
};
const componentToProps = (
  jsDocs: JSDocTypes,
  component: JSDocType,
  typeProp: JSDocType,
): JSDocType[] => {
  if (typeProp.parameters?.length) {
    return walkProps(jsDocs, typeProp.parameters[0]);
  }
  return [];
};
export const extractProps: FrameworkPlugin = parse => {
  return Object.keys(parse.structures).reduce((acc, name) => {
    const component = parse.structures[name];
    const typeProp = reactComponent(component, parse.imports);
    if (typeProp) {
      const reactComponent = { ...component };
      reactComponent.properties = componentToProps(
        parse.structures,
        component,
        typeProp,
      );
      delete reactComponent.parameters;
      delete reactComponent.inherits;
      delete reactComponent.value;
      return { ...acc, [name]: reactComponent };
    }
    return { ...acc, [name]: component };
  }, {});
};
