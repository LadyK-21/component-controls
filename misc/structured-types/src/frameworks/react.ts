import * as ts from 'typescript';
import { getSymbolType, getFunctionLike, getInitializer } from '../ts-utils';
import {
  isObjectTypeDeclaration,
  getObjectStaticProp,
  ParsePlugin,
} from '../ts-utils';
import { PropKind } from '../types';

const reactPLugin: ParsePlugin = {
  internalTypes: ['PropsWithChildren'],
  typesResolve: ({ symbolType, declaration, checker }) => {
    let propsType = symbolType;
    let initializer: ts.Node | undefined = undefined;
    let name = undefined;
    let kind = undefined;
    let framework = undefined;
    let defaultProps = undefined;
    let displayName = undefined;
    if ((symbolType.flags & ts.TypeFlags.Object) === ts.TypeFlags.Object) {
      if (declaration) {
        if (isObjectTypeDeclaration(declaration)) {
          initializer = getInitializer(declaration);
          const jsx = checker.getJsxIntrinsicTagNamesAt(declaration);
          if (jsx) {
            defaultProps = getObjectStaticProp(declaration, 'defaultProps');
            displayName = getObjectStaticProp(declaration, 'displayName');

            framework = 'react';
            kind = PropKind.Class;
            const signatures = symbolType.getConstructSignatures();
            if (signatures.length > 0 && signatures[0].parameters.length) {
              const props = signatures[0].parameters[0];
              propsType = getSymbolType(checker, props) || symbolType;

              if (propsType.isUnionOrIntersection()) {
                propsType = propsType.types[0];
              }
            }
          }
        } else {
          const reactFunction = getFunctionLike(checker, declaration);
          if (reactFunction) {
            const jsx = checker.getJsxIntrinsicTagNamesAt(reactFunction);
            if (jsx) {
              defaultProps = getObjectStaticProp(
                reactFunction.parent,
                'defaultProps',
              );
              displayName =
                getObjectStaticProp(reactFunction.parent, 'displayName') ||
                reactFunction.name?.getText() ||
                (ts.isVariableDeclaration(reactFunction.parent) &&
                  reactFunction.parent.name.getText());
              framework = 'react';
              kind = PropKind.Function;
              if (reactFunction.parameters.length) {
                const props = reactFunction.parameters[0];
                if (ts.isObjectBindingPattern(props.name)) {
                  initializer = props.name;
                }
                propsType = checker.getTypeAtLocation(props);
              }
            }
          }
        }
      }
    }
    name =
      typeof displayName === 'string'
        ? displayName
        : displayName && ts.isStringLiteral(displayName)
        ? displayName.text
        : undefined;
    if (defaultProps) {
      initializer = defaultProps;
    }

    return {
      type: propsType,
      name,
      initializer,
      kind,
      framework,
      skipGenerics: kind === PropKind.Function,
      skipParameters: kind === PropKind.Function,
    };
  },
};

export default reactPLugin;
