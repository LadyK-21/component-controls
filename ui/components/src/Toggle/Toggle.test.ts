import path from 'path';
import MatchMediaMock from 'jest-matchmedia-mock';
import { loadConfigurations } from '@component-controls/config';
import { renderExample } from '@component-controls/test-renderers';
import { render as reactRender } from '@component-controls/render/react';
import { render, act } from '@testing-library/react';

const renderErr = () => {
  throw new Error('Could not render the story');
};

import doc, {
  overview,
  label,
  customIcons,
  octicons,
  checkbox,
} from './Toggle.stories';

describe('Toggle', () => {
  let matchMedia: MatchMediaMock;
  beforeAll(() => {
    jest.mock('rc-util/lib/Portal');
    matchMedia = new MatchMediaMock();
  });
  afterEach(() => {
    matchMedia.clear();
  });
  const configPath = path.resolve(__dirname, '../../.config');
  const config = loadConfigurations(configPath);
  if (!config.renderFn) {
    config.renderFn = reactRender;
  }

  test('overview', () => {
    const example = overview;

    let rendered;
    act(() => {
      rendered = renderExample({
        example,
        doc,
        config,
      });
    });
    if (!rendered) {
      renderErr();
      return;
    }
    const { asFragment } = render(rendered);
    expect(asFragment()).toMatchSnapshot();
  });

  test('label', () => {
    const example = label;

    let rendered;
    act(() => {
      rendered = renderExample({
        example,
        doc,
        config,
      });
    });
    if (!rendered) {
      renderErr();
      return;
    }
    const { asFragment } = render(rendered);
    expect(asFragment()).toMatchSnapshot();
  });

  test('customIcons', () => {
    const example = customIcons;

    let rendered;
    act(() => {
      rendered = renderExample({
        example,
        doc,
        config,
      });
    });
    if (!rendered) {
      renderErr();
      return;
    }
    const { asFragment } = render(rendered);
    expect(asFragment()).toMatchSnapshot();
  });

  test('octicons', () => {
    const example = octicons;

    let rendered;
    act(() => {
      rendered = renderExample({
        example,
        doc,
        config,
      });
    });
    if (!rendered) {
      renderErr();
      return;
    }
    const { asFragment } = render(rendered);
    expect(asFragment()).toMatchSnapshot();
  });

  test('checkbox', () => {
    const example = checkbox;

    let rendered;
    act(() => {
      rendered = renderExample({
        example,
        doc,
        config,
      });
    });
    if (!rendered) {
      renderErr();
      return;
    }
    const { asFragment } = render(rendered);
    expect(asFragment()).toMatchSnapshot();
  });
});
