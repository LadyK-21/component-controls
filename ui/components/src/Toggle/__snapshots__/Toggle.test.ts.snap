// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle checkbox 1`] = `
<DocumentFragment>
  <label
    class="css-knfhfe"
  >
    <div
      class="css-x3crln"
    >
      <input
        checked=""
        class="css-p19g2b"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="css-1dq90k0"
        fill="currentcolor"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="css-1b36guq"
        fill="currentcolor"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </div>
    theme-ui checkbox
  </label>
</DocumentFragment>
`;

exports[`Toggle customIcons 1`] = `
<DocumentFragment>
  <div
    style="position: relative; display: inline-block; text-align: left; opacity: 1; direction: ltr; border-radius: 14px; transition: opacity 0.25s; user-select: none;"
  >
    <div
      class="react-switch-bg"
      style="height: 28px; width: 56px; margin: 0px; position: relative; background: rgb(136, 136, 136); border-radius: 14px; cursor: pointer; transition: background 0.25s;"
    >
      <div
        style="height: 28px; width: 30px; position: relative; opacity: 0; pointer-events: none; transition: opacity 0.25s;"
      >
        <svg
          fill="aqua"
          height="100%"
          viewBox="0 0 10 10"
          width="100%"
        >
          <circle
            cx="5"
            cy="5"
            r="3"
          />
        </svg>
      </div>
      <div
        style="height: 28px; width: 30px; position: absolute; opacity: 1; right: 0px; top: 0px; pointer-events: none; transition: opacity 0.25s;"
      >
        <svg
          fill="red"
          height="100%"
          viewBox="0 0 10 10"
          width="100%"
        >
          <circle
            cx="5"
            cy="5"
            r="3"
          />
        </svg>
      </div>
    </div>
    <div
      class="react-switch-handle"
      style="height: 26px; width: 26px; background: rgb(255, 255, 255); display: inline-block; cursor: pointer; border-radius: 50%; position: absolute; transform: translateX(1px); top: 1px; outline: 0; border: 0px; transition: background-color 0.25s, transform 0.25s, box-shadow 0.15s;"
    />
    <input
      aria-checked="false"
      role="switch"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px;"
      type="checkbox"
    />
  </div>
</DocumentFragment>
`;

exports[`Toggle label 1`] = `
<DocumentFragment>
  <label
    class="css-knfhfe"
    for="custom-toggle"
  >
    <span
      class="css-1iu2cu0-Toggle"
    >
      with label
    </span>
    <div
      style="position: relative; display: inline-block; text-align: left; opacity: 1; direction: ltr; border-radius: 14px; transition: opacity 0.25s; user-select: none;"
    >
      <div
        class="react-switch-bg"
        style="height: 28px; width: 56px; margin: 0px; position: relative; background: rgb(136, 136, 136); border-radius: 14px; cursor: pointer; transition: background 0.25s;"
      >
        <div
          style="height: 28px; width: 30px; position: relative; opacity: 0; pointer-events: none; transition: opacity 0.25s;"
        >
          <svg
            height="100%"
            style="position: absolute; top: 0px;"
            viewBox="-2 -5 17 21"
            width="100%"
          >
            <path
              d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
              fill="#fff"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          style="height: 28px; width: 30px; position: absolute; opacity: 1; right: 0px; top: 0px; pointer-events: none; transition: opacity 0.25s;"
        >
          <svg
            height="100%"
            style="position: absolute; top: 0px;"
            viewBox="-2 -5 14 20"
            width="100%"
          >
            <path
              d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
              fill="#fff"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
      <div
        class="react-switch-handle"
        style="height: 26px; width: 26px; background: rgb(255, 255, 255); display: inline-block; cursor: pointer; border-radius: 50%; position: absolute; transform: translateX(1px); top: 1px; outline: 0; border: 0px; transition: background-color 0.25s, transform 0.25s, box-shadow 0.15s;"
      />
      <input
        aria-checked="false"
        id="custom-toggle"
        role="switch"
        style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px;"
        type="checkbox"
      />
    </div>
  </label>
</DocumentFragment>
`;

exports[`Toggle octicons 1`] = `
<DocumentFragment>
  <div
    style="position: relative; display: inline-block; text-align: left; opacity: 1; direction: ltr; border-radius: 10px; transition: opacity 0.25s; user-select: none;"
  >
    <div
      class="react-switch-bg"
      style="height: 20px; width: 40px; margin: 0px; position: relative; background: rgb(211, 211, 211); border-radius: 10px; cursor: pointer; transition: background 0.25s;"
    >
      <div
        style="height: 20px; width: 24px; position: relative; opacity: 0; pointer-events: none; transition: opacity 0.25s;"
      >
        <div
          class="css-7kjknt"
        >
          <div
            class="css-r0wgrp"
          >
            <svg
              aria-hidden="true"
              class="octicon"
              fill="currentColor"
              height="16"
              role="img"
              style="display: inline-block; user-select: none; vertical-align: text-bottom;"
              viewBox="0 0 16 16"
              width="16"
            >
              <path
                d="M1.679 7.932c.412-.621 1.242-1.75 2.366-2.717C5.175 4.242 6.527 3.5 8 3.5c1.473 0 2.824.742 3.955 1.715 1.124.967 1.954 2.096 2.366 2.717a.119.119 0 010 .136c-.412.621-1.242 1.75-2.366 2.717C10.825 11.758 9.473 12.5 8 12.5c-1.473 0-2.824-.742-3.955-1.715C2.92 9.818 2.09 8.69 1.679 8.068a.119.119 0 010-.136zM8 2c-1.981 0-3.67.992-4.933 2.078C1.797 5.169.88 6.423.43 7.1a1.619 1.619 0 000 1.798c.45.678 1.367 1.932 2.637 3.024C4.329 13.008 6.019 14 8 14c1.981 0 3.67-.992 4.933-2.078 1.27-1.091 2.187-2.345 2.637-3.023a1.619 1.619 0 000-1.798c-.45-.678-1.367-1.932-2.637-3.023C11.671 2.992 9.981 2 8 2zm0 8a2 2 0 100-4 2 2 0 000 4z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        style="height: 20px; width: 24px; position: absolute; opacity: 1; right: 0px; top: 0px; pointer-events: none; transition: opacity 0.25s;"
      >
        <div
          class="css-1lbnzn6"
        >
          <div
            class="css-1jfm6h9"
          >
            <svg
              aria-hidden="true"
              class="octicon"
              fill="currentColor"
              height="16"
              role="img"
              style="display: inline-block; user-select: none; vertical-align: text-bottom;"
              viewBox="0 0 16 16"
              width="16"
            >
              <path
                d="M.143 2.31a.75.75 0 011.047-.167l14.5 10.5a.75.75 0 11-.88 1.214l-2.248-1.628C11.346 13.19 9.792 14 8 14c-1.981 0-3.67-.992-4.933-2.078C1.797 10.832.88 9.577.43 8.9a1.618 1.618 0 010-1.797c.353-.533.995-1.42 1.868-2.305L.31 3.357A.75.75 0 01.143 2.31zm3.386 3.378a14.21 14.21 0 00-1.85 2.244.12.12 0 00-.022.068c0 .021.006.045.022.068.412.621 1.242 1.75 2.366 2.717C5.175 11.758 6.527 12.5 8 12.5c1.195 0 2.31-.488 3.29-1.191L9.063 9.695A2 2 0 016.058 7.52l-2.53-1.832zM8 3.5c-.516 0-1.017.09-1.499.251a.75.75 0 11-.473-1.423A6.23 6.23 0 018 2c1.981 0 3.67.992 4.933 2.078 1.27 1.091 2.187 2.345 2.637 3.023a1.619 1.619 0 010 1.798c-.11.166-.248.365-.41.587a.75.75 0 11-1.21-.887c.148-.201.272-.382.371-.53a.119.119 0 000-.137c-.412-.621-1.242-1.75-2.366-2.717C10.825 4.242 9.473 3.5 8 3.5z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="react-switch-handle"
      style="height: 14px; width: 14px; background: rgb(255, 255, 255); display: inline-block; cursor: pointer; border-radius: 50%; position: absolute; transform: translateX(3px); top: 3px; outline: 0; border: 0px; transition: background-color 0.25s, transform 0.25s, box-shadow 0.15s;"
    />
    <input
      aria-checked="false"
      role="switch"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px;"
      type="checkbox"
    />
  </div>
</DocumentFragment>
`;

exports[`Toggle overview 1`] = `
<DocumentFragment>
  <div
    style="position: relative; display: inline-block; text-align: left; opacity: 1; direction: ltr; border-radius: 14px; transition: opacity 0.25s; user-select: none;"
  >
    <div
      class="react-switch-bg"
      style="height: 28px; width: 56px; margin: 0px; position: relative; background: rgb(136, 136, 136); border-radius: 14px; cursor: pointer; transition: background 0.25s;"
    >
      <div
        style="height: 28px; width: 30px; position: relative; opacity: 0; pointer-events: none; transition: opacity 0.25s;"
      >
        <svg
          height="100%"
          style="position: absolute; top: 0px;"
          viewBox="-2 -5 17 21"
          width="100%"
        >
          <path
            d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
            fill="#fff"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        style="height: 28px; width: 30px; position: absolute; opacity: 1; right: 0px; top: 0px; pointer-events: none; transition: opacity 0.25s;"
      >
        <svg
          height="100%"
          style="position: absolute; top: 0px;"
          viewBox="-2 -5 14 20"
          width="100%"
        >
          <path
            d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
            fill="#fff"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      class="react-switch-handle"
      style="height: 26px; width: 26px; background: rgb(255, 255, 255); display: inline-block; cursor: pointer; border-radius: 50%; position: absolute; transform: translateX(1px); top: 1px; outline: 0; border: 0px; transition: background-color 0.25s, transform 0.25s, box-shadow 0.15s;"
    />
    <input
      aria-checked="false"
      role="switch"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px;"
      type="checkbox"
    />
  </div>
</DocumentFragment>
`;
